VKV version 0.7.1 and later suddenly fail with "[ERROR] not authenticated..." Worked on 0.7.0.

Hello,

With vkv versions 0.7.1 and later I get an error on some operations like 'vkv list engines' and 'vkv export'.

### Some background
- we tested this setup on an ARM64 Mac using Minikube
- vkv binaries used are therefore also Darwin arm64 binaries
- errors reproduced on Vault versions 1.16.3 AND 1.18.0
- engine tested is Vault KV2 engine
- we use multiple Vault instances in multiple namespaces so naming scheme for engine mount points is "<k8s-namespace>/<name-of-secret-engine>"
- in this case, on Minikube namespace 'default', the KV2 engine 'secrets' is mounted on mountpoint 'default/secrets'

### Mountpoint info Vault
```
vault 2024-10-28T10:49:32.364Z [INFO]  core: successfully mounted: type=system version="v1.18.0+builtin.vault" path=sys/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.365Z [INFO]  core: successfully mounted: type=identity version="v1.18.0+builtin.vault" path=identity/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.365Z [INFO]  core: successfully mounted: type=cubbyhole version="v1.18.0+builtin.vault" path=cubbyhole/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.365Z [INFO]  core: successfully mounted: type=kv version="v0.20.0+builtin" path=default/secrets/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.367Z [INFO]  core: successfully mounted: type=token version="v1.18.0+builtin.vault" path=token/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.367Z [INFO]  core: successfully mounted: type=userpass version="v1.18.0+builtin.vault" path=default/userpass/ namespace="ID: root. Path: "
vault 2024-10-28T10:49:32.368Z [INFO]  core: successfully mounted: type=approle version="v1.18.0+builtin.vault" path=default/approle/ namespace="ID: root. Path: "
```

### How to reproduce
In this testsetup I have the VAULT_TOKEN env var set, containing the ROOT TOKEN of the targeted Vault instance.

Also I have the following additional env vars set:
```
$ printenv | grep ^VKV
VKV_EXPORT_FORMAT=json
VKV_LIST_ENGINES_FORMAT=json
VKV_EXPORT_ENGINE_PATH=default/secrets

$ printenv | grep ^VAULT
VAULT_FORMAT=json
VAULT_ADDR=https://xxx-vault.xxx.localdev
VAULT_TOKEN=hvs.xxxxxxxxxx
VAULT_SKIP_VERIFY=true
```

The vkv version 0.7.0 outputs the following:
```
$ vkv-0.7.0 version
vkv 0.7.0

$ vkv-0.7.0 list engines
{
  "engines": [
    "default/secrets/"
  ]
}
$ vkv-0.7.0 export 
{
  "default/secrets": {
    "foo/": {
      "bar/": {
        "testkey2": {
          "value": "testvalue2"
        }
      },
      "testkey1": {
        "value": "testvalue1"
      }
    }
(...)
```


The vkv version 0.7.1 and later all output like the following:
```
$ vkv-0.7.1 version
vkv 0.7.1

$ vkv-0.7.1 list engines
[ERROR] not authenticated, perhaps not a valid token: Error making API request.

URL: GET https://xxx-vault.xxx.localdev/v1/auth/token/lookup-self
Code: 403. Errors:

* 2 errors occurred:
        * permission denied
        * invalid token


$ vkv-0.7.1 export 
[ERROR] not authenticated, perhaps not a valid token: Error making API request.

URL: GET https://xxx-vault.xxx.localdev/v1/auth/token/lookup-self
Code: 403. Errors:

* 2 errors occurred:
        * permission denied
        * invalid token

.
```

Could not find any clues in web documentation / 'vkv --help' output.

Hope you can help / advise on this matter.

Best regards,

Emiel Kuiper

--- a/.golang-ci.yml
+++ b/.golang-ci.yml
@@ -28,3 +28,6 @@ linters:
     - revive
     - depguard
     - tagalign
+    - copyloopvar
+    - intrange
+    - execinquery

--- a/cmd/root_test.go
+++ b/cmd/root_test.go
@@ -118,7 +118,7 @@ func (s *VaultSuite) TestMode() {
 
 func TestVaultSuite(t *testing.T) {
 	// github actions doesn't offer the docker socket, which we need to run this test suite
-	if runtime.GOOS == "linux" {
+	if runtime.GOOS != "windows" {
 		suite.Run(t, new(VaultSuite))
 	}
 }

--- a/docs/authentication.md
+++ b/docs/authentication.md
@@ -2,7 +2,7 @@
 
 `vkv` supports all of Vaults [environment variables](https://www.vaultproject.io/docs/commands#environment-variables) as well as any configured [Token helpers](https://developer.hashicorp.com/vault/docs/commands/token-helper). 
 
-In order to authenticate you will have to set at least `VAULT_ADDR` and `VAULT_TOKEN`.
+In order to authenticate you will have to set at least one of the `VAULT_ADDR` or `VKV_LOGIN_COMMAND` and `VAULT_TOKEN` env vars.
 
 ## MacOS/Linux
 ```
@@ -20,7 +20,7 @@ vkv.exe export --path <KVv2-path>
 
 ## Special Env Var `VKV_LOGIN_COMMAND`
 For advanced use cases, you can set `VKV_LOGIN_COMMAND`, that way `vkv` will first execute the specified command and use the output of the command as the token.
-This is way you dont have to hardcode and set `VAULT_TOKEN`, this is especially useful when using `vkv` in CI. (See Gitlab Integration):
+This is way you don't have to hardcode and set `VAULT_TOKEN`, this is especially useful when using `vkv` in CI. (See Gitlab Integration):
 
 Example:
 
@@ -28,3 +28,25 @@ Example:
 export VKV_LOGIN_COMMAND="vault write -field=token auth/jwt/login jwt=${CI_JOB_JWT_V2}"
 vkv export -p
 ```
+
+## Token Precedence
+The following token precedence is applied (from highest to lowest):
+
+1. `VKV_TOKEN`
+2. `VKV_LOGIN_COMMAND`
+3. [Vault Token Helper](https://developer.hashicorp.com/vault/docs/commands/token-helper), where the token will be written to `~/.vault-token`.
+
+If `vkv` detects **more than one possible token source**, warnings are shown as the following, indicating which token source will be used:
+
+```bash
+$> vkv export -p secret
+[WARN] More than one token source configured (either VAULT_TOKEN, VKV_LOGIN_COMMAND or ~/.vault-token).
+[WARN] See https://falcosuessgott.github.io/vkv/authentication/#token-precedence for vkv's token precedence logic. Disable these warnings with VKV_DISABLE_WARNING.
+[INFO] Using VAULT_TOKEN.
+
+secret/ [desc=key/value secret storage] [type=kv2]
+└── secret [v=1]
+    └── key=*****
+```
+
+As described, one can disable these warning by setting `VKV_DISABLE_WARNING` to any value.

--- a/pkg/testutils/testutils_test.go
+++ b/pkg/testutils/testutils_test.go
@@ -52,7 +52,7 @@ func (s *VaultSuite) TestVaultConnection() {
 
 func TestVaultSuite(t *testing.T) {
 	// github actions doesn't offer the docker socket, which we need to run this test suite
-	if runtime.GOOS == "linux" {
+	if runtime.GOOS != "windows" {
 		suite.Run(t, new(VaultSuite))
 	}
 }

--- a/pkg/vault/client.go
+++ b/pkg/vault/client.go
@@ -18,45 +18,18 @@ type Vault struct {
 
 // NewDefaultClient returns a new vault client wrapper.
 func NewDefaultClient() (*Vault, error) {
+	token, err := getToken()
+	if err != nil {
+		return nil, err
+	}
+
 	// create vault client using defaults (recommended)
 	c, err := api.NewClient(nil)
 	if err != nil {
 		return nil, err
 	}
 
-	// use tokenhelper if available
-	th, err := tokenhelper.NewInternalTokenHelper()
-	if err != nil {
-		return nil, fmt.Errorf("error creating default token helper: %w", err)
-	}
-
-	token, err := th.Get()
-	if err != nil {
-		return nil, fmt.Errorf("error getting token from default token helper: %w", err)
-	}
-
-	if token != "" {
-		c.SetToken(token)
-	}
-
-	// custom: if VKV_LOGIN_COMMAND is set, execute it and set the output as token
-	cmd, ok := os.LookupEnv("VKV_LOGIN_COMMAND")
-	if ok && cmd != "" {
-		cmdParts := strings.Split(cmd, " ")
-
-		token, err := exec.Run(cmdParts)
-		if err != nil {
-			return nil, fmt.Errorf("error running VKV_LOGIN_CMD (%s): %w", cmd, err)
-		}
-
-		vaultToken := strings.TrimSpace(string(token))
-		if vaultToken == "" {
-			return nil, errors.New("VKV_LOGIN_COMMAND required but not set")
-		}
-
-		// set token
-		c.SetToken(vaultToken)
-	}
+	c.SetToken(token)
 
 	// self lookup current auth for verification
 	if _, err := c.Auth().Token().LookupSelf(); err != nil {
@@ -80,4 +53,103 @@ func NewClient(addr, token string) (*Vault, error) {
 	c.SetToken(token)
 
 	return &Vault{Client: c}, nil
+}
+
+// getToken finds the token configured by the user via env vars or token helpers
+// Precedence: 1. VAULT_TOKEN, 2. VKV_LOGIN_COMMAND, 3. Vault Token Helper.
+//
+//nolint:cyclop
+func getToken() (string, error) {
+	// warn user if more than one is configured
+	envToken, envTokenOk := os.LookupEnv("VAULT_TOKEN")
+	tokenCommand, tokenCommandOk := os.LookupEnv("VKV_LOGIN_COMMAND")
+
+	th, err := tokenhelper.NewInternalTokenHelper()
+	if err != nil {
+		return "", fmt.Errorf("error creating default token helper: %w", err)
+	}
+
+	thToken, err := th.Get()
+	if err != nil {
+		return "", fmt.Errorf("error getting token from default token helper: %w", err)
+	}
+
+	var (
+		// number of tokens configured
+		tokenSources int
+
+		// if we issue a warning to the user, we also want to inform with what token option we went
+		warn bool
+	)
+
+	if envTokenOk {
+		tokenSources++
+	}
+
+	if tokenCommandOk {
+		tokenSources++
+	}
+
+	if thToken != "" {
+		tokenSources++
+	}
+
+	// check whether user disabled warnings
+	_, disableWarn := os.LookupEnv("VKV_DISABLE_WARNING")
+
+	if tokenSources > 1 {
+		warn = true
+
+		if !disableWarn {
+			fmt.Println("[WARN] More than one token source configured (either VAULT_TOKEN, VKV_LOGIN_COMMAND or ~/.vault-token).")
+			fmt.Println("[WARN] See https://falcosuessgott.github.io/vkv/authentication/token-precedence for vkv's token precedence logic. Disable these warnings with VKV_DISABLE_WARNING.")
+		}
+	}
+
+	// if VAULT_TOKEN is set - return it
+	if envToken != "" {
+		if warn && !disableWarn {
+			fmt.Println("[INFO] Using VAULT_TOKEN.")
+			fmt.Println()
+		}
+
+		return envToken, nil
+	}
+
+	// if VKV_LOGIN_COMMAND
+	if tokenCommand != "" {
+		if warn && !disableWarn {
+			fmt.Println("[INFO] Using VKV_LOGIN_COMMAND.")
+			fmt.Println()
+		}
+
+		return runVaultTokenCommand(tokenCommand)
+	}
+
+	if thToken != "" {
+		if warn && !disableWarn {
+			fmt.Println("[INFO] Using ~/.vault-token.")
+			fmt.Println()
+		}
+
+		return thToken, nil
+	}
+
+	return "", errors.New("no token provided")
+}
+
+func runVaultTokenCommand(cmd string) (string, error) {
+	cmdParts := strings.Split(cmd, " ")
+
+	token, err := exec.Run(cmdParts)
+	if err != nil {
+		return "", fmt.Errorf("error running VKV_LOGIN_CMD (%s): %w", cmd, err)
+	}
+
+	vaultToken := strings.TrimSpace(string(token))
+	if vaultToken == "" {
+		return "", errors.New("VKV_LOGIN_COMMAND required but not set")
+	}
+
+	return vaultToken, nil
 }

--- a/pkg/vault/client_test.go
+++ b/pkg/vault/client_test.go
@@ -39,6 +39,57 @@ func (s *VaultSuite) SetupSubTest() {
 	s.client = v
 }
 
+func (s *VaultSuite) TestGetToken() {
+	testCases := []struct {
+		name     string
+		envVars  map[string]string
+		expToken string
+		err      bool
+	}{
+		{
+			name:     "vault token",
+			expToken: "token",
+			envVars: map[string]string{
+				"VAULT_TOKEN": "token",
+			},
+		},
+		{
+			name:     "vkv login command",
+			expToken: "testtoken",
+			envVars: map[string]string{
+				"VKV_LOGIN_COMMAND": "echo testtoken",
+			},
+		},
+		{
+			name: "none",
+			err:  true,
+		},
+	}
+
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			// unsetting any local VAULT_TOKEN env var
+			os.Unsetenv("VAULT_TOKEN")
+
+			// set env vars
+			for k, v := range tc.envVars {
+				s.T().Setenv(k, v)
+			}
+
+			// invoke token
+			t, err := getToken()
+
+			// assert
+			if tc.err {
+				s.Require().Error(err, tc.name)
+			} else {
+				s.Require().NoError(err, tc.name)
+				s.Require().Equal(tc.expToken, t, tc.name)
+			}
+		})
+	}
+}
+
 func (s *VaultSuite) TestNewClient() {
 	testCases := []struct {
 		name                           string
@@ -120,7 +171,7 @@ func (s *VaultSuite) TestNewClient() {
 func TestVaultSuite(t *testing.T) {
 	// github actions doenst offer the docker sock, which we need
 	// to run this test suite
-	if runtime.GOOS == "linux" {
+	if runtime.GOOS != "windows" {
 		suite.Run(t, new(VaultSuite))
 	}
 }

